### PR TITLE
Open channel

### PR DESCRIPTION
**Pull request by holzeis**
see fb629ef63f56ac498a03cef6fb8634cfc32d02c4

### CODE REVIEW

**Comment by holzeis** [comment]
Somehow the channels don't get opened anymore, although I was sure it was working after 1677a5269a26674a7d1de753a019fe01743b6880.

I think the following log message on the counterparty may be the reason. The taker node is not reachable from the counterparty. Although I am not sure why this is the case in my local network (it is realistic though in a real world scenario)
```
2022-11-12 07:45:42.943 DEBUG [lightning::ln::channelmanager:6061] Marking channels with 038a567aefaeba8cca6156ed1b4e46038af2138929a8ced9e86dcfb539be0ee689 disconnected and generating channel_updates. We believe we can make future connections to this peer.
```

**Comment by holzeis** [comment]
Some notes from my debugging.

- ldk-bdk sample node to another ldk-bdk sample node seems to be working, however I've never seen an onchain transaction supporting the opened channel.
- It looks like the taker node (mobile device) is not reachable from the counter party. Although I am not sure why this is the case in my local network (it is realistic though in a real world scenario) - It might be the reason though why the channel does not get opened through our app.
```
2022-11-12 07:45:42.943 DEBUG [lightning::ln::channelmanager:6061] Marking channels with 038a567aefaeba8cca6156ed1b4e46038af2138929a8ced9e86dcfb539be0ee689 disconnected and generating channel_updates. We believe we can make future connections to this peer.
```
- Peer connection from the taker gets listed on the counter party node - so the connection is definitely established.

**Comment by klochowicz** [comment]
would you mind splitting this PR up? the first commit can probably go in as-is to fix our problems (and then we can maybe we could pair up to align on `Balance` changes ?)

**Comment by holzeis** [comment]
One step close d95c95655e7737afda0046b01c1c0308ec6d8e06. thread handles got dropped immediately after init wallet function completed, hence no incoming messages could be processed (tcp handle) and the background processor (also starts a thread) did not process any events, etc.

However opening a channel soley through the app is still not working as the counter party (maker) is not able to connect to the phone (taker) upon channel creation. see (maker) logs below.

```
2022-11-14 10:50:58.027 DEBUG [lightning::ln::peer_handler:1111] Finished noise handshake for connection with 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:50:58.027 TRACE [lightning::ln::peer_handler:980] Enqueueing message Init { features: [162, 81, 10, 8, 128, 160, 8], remote_network_address: None } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:50:58.029 INFO  [lightning::ln::peer_handler:1239] Received peer Init message from 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d: DataLossProtect: supported, InitialRoutingSync: not supported, UpfrontShutdownScript: supported, GossipQueries: supported, VariableLengthOnion: required, StaticRemoteKey: required, PaymentSecret: required, BasicMPP: supported, Wumbo: supported, ShutdownAnySegwit: supported, OnionMessages: supported, ChannelType: supported, SCIDPrivacy: supported, ZeroConf: supported, unknown flags: none
2022-11-14 10:50:58.029 DEBUG [lightning::ln::channelmanager:6116] Generating channel_reestablish events for 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:50:58.030 TRACE [lightning_background_processor:520] Persisting ChannelManager...
2022-11-14 10:50:58.031 TRACE [lightning::ln::peer_handler:980] Enqueueing message GossipTimestampFilter { chain_hash: 000000000933ea01ad0ee984209779baaec3ced90fa3f408719526f8d77f4943, first_timestamp: 1667213458, timestamp_range: 4294967295 } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:50:58.069 TRACE [lightning_background_processor:520] Done persisting ChannelManager.
2022-11-14 10:51:05.846 DEBUG [lightning::ln::peer_handler:2035] Broadcasting NodeAnnouncement after passing it to our own RoutingMessageHandler.
2022-11-14 10:51:07.712 TRACE [lightning::ln::channelmanager:5716] New best block: 00000000000059fdf1f7475c15558296210223624395c5144d0e46ac202c7e3b at height 2406224
2022-11-14 10:51:07.713 TRACE [lightning_background_processor:520] Persisting ChannelManager...
2022-11-14 10:51:07.713 DEBUG [lightning::chain::chainmonitor:537] New best block 00000000000059fdf1f7475c15558296210223624395c5144d0e46ac202c7e3b at height 2406224 provided via best_block_updated
2022-11-14 10:51:07.733 TRACE [lightning_background_processor:520] Done persisting ChannelManager.
2022-11-14 10:51:12.110 TRACE [lightning_background_processor:520] Calling ChannelManager's timer_tick_occurred
2022-11-14 10:51:16.301 TRACE [lightning_background_processor:520] Calling PeerManager's timer_tick_occurred
2022-11-14 10:51:16.301 TRACE [lightning::ln::peer_handler:980] Enqueueing message Ping { ponglen: 0, byteslen: 64 } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:51:16.716 TRACE [lightning_background_processor:520] Persisting scorer
2022-11-14 10:51:37.696 TRACE [lightning::ln::channelmanager:5716] New best block: 00000000000059fdf1f7475c15558296210223624395c5144d0e46ac202c7e3b at height 2406224
2022-11-14 10:51:37.698 TRACE [lightning_background_processor:520] Persisting ChannelManager...
2022-11-14 10:51:37.698 DEBUG [lightning::chain::chainmonitor:537] New best block 00000000000059fdf1f7475c15558296210223624395c5144d0e46ac202c7e3b at height 2406224 provided via best_block_updated
2022-11-14 10:51:37.724 TRACE [lightning_background_processor:520] Done persisting ChannelManager.
2022-11-14 10:51:45.486 TRACE [lightning::ln::peer_handler:1835] Handling disconnection of peer 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d, with future connection to the peer possible.
2022-11-14 10:51:45.486 DEBUG [lightning::ln::channelmanager:6061] Marking channels with 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d disconnected and generating channel_updates. We believe we can make future connections to this peer.
```

If however, I initiate a connection from the counter party (maker) to the phone (taker)

```
> connectpeer 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d@127.0.0.1:9735
SUCCESS: connected to peer 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
> listpeers
	{
		 pubkey: 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
	},
```

and create a channel afterward through the 10101 app. The channel gets opened successfully.

```
> listchannels
[
	{
		channel_id: 371a1657559bcc1e92e2876a08d18daa7f8216821c5c86990d3da1e95f60420e,
		funding_txid: 0e42605fe9a13d0d99865c1c8216827faa8dd1086a87e2921ecc9b5557161a37,
		peer_pubkey: 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d,
		is_channel_ready: false,
		channel_value_satoshis: 21000,
		local_balance_msat: 0,
		channel_can_send_payments: false,
		public: false,
	},
]
```

see (maker) logs below

```
2022-11-14 10:58:47.011 TRACE [lightning::ln::peer_handler:980] Enqueueing message Init { features: [162, 81, 10, 8, 128, 160, 8], remote_network_address: None } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:58:47.013 INFO  [lightning::ln::peer_handler:1239] Received peer Init message from 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d: DataLossProtect: supported, InitialRoutingSync: not supported, UpfrontShutdownScript: supported, GossipQueries: supported, VariableLengthOnion: required, StaticRemoteKey: required, PaymentSecret: required, BasicMPP: supported, Wumbo: supported, ShutdownAnySegwit: supported, OnionMessages: supported, ChannelType: supported, SCIDPrivacy: supported, ZeroConf: supported, unknown flags: none
2022-11-14 10:58:47.014 DEBUG [lightning::ln::channelmanager:6116] Generating channel_reestablish events for 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:58:47.014 TRACE [lightning_background_processor:520] Persisting ChannelManager...
2022-11-14 10:58:47.014 TRACE [lightning::ln::peer_handler:980] Enqueueing message GossipTimestampFilter { chain_hash: 000000000933ea01ad0ee984209779baaec3ced90fa3f408719526f8d77f4943, first_timestamp: 1667213927, timestamp_range: 4294967295 } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:58:47.068 TRACE [lightning_background_processor:520] Done persisting ChannelManager.
2022-11-14 10:59:02.232 TRACE [lightning_background_processor:520] Calling PeerManager's timer_tick_occurred
2022-11-14 10:59:02.232 TRACE [lightning::ln::peer_handler:980] Enqueueing message Ping { ponglen: 0, byteslen: 64 } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:59:02.241 TRACE [lightning::ln::peer_handler:1283] Received message Pong(Pong { byteslen: 0 }) from 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:59:02.857 TRACE [lightning_background_processor:520] Persisting scorer
2022-11-14 10:59:05.851 DEBUG [lightning::ln::peer_handler:2035] Broadcasting NodeAnnouncement after passing it to our own RoutingMessageHandler.
2022-11-14 10:59:07.699 TRACE [lightning::ln::channelmanager:5716] New best block: 000000000000004477a4cf53f2f27ef34a1cf5b0cc396a9564148e2121f41085 at height 2406225
2022-11-14 10:59:07.700 DEBUG [lightning::chain::chainmonitor:537] New best block 000000000000004477a4cf53f2f27ef34a1cf5b0cc396a9564148e2121f41085 at height 2406225 provided via best_block_updated
2022-11-14 10:59:07.700 TRACE [lightning_background_processor:520] Persisting ChannelManager...
2022-11-14 10:59:07.721 TRACE [lightning_background_processor:520] Done persisting ChannelManager.
2022-11-14 10:59:09.466 TRACE [lightning::ln::peer_handler:1283] Received message Ping(Ping { ponglen: 0, byteslen: 64 }) from 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:59:09.467 TRACE [lightning::ln::peer_handler:980] Enqueueing message Pong { byteslen: 0 } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:59:21.654 TRACE [lightning_background_processor:520] Calling ChannelManager's timer_tick_occurred
2022-11-14 10:59:32.113 TRACE [lightning::ln::peer_handler:1283] Received message OpenChannel(OpenChannel { chain_hash: 000000000933ea01ad0ee984209779baaec3ced90fa3f408719526f8d77f4943, temporary_channel_id: [193, 82, 42, 154, 220, 195, 230, 191, 254, 94, 154, 29, 95, 37, 182, 32, 67, 204, 167, 49, 233, 216, 49, 107, 164, 88, 245, 206, 10, 52, 168, 3], funding_satoshis: 21000, push_msat: 0, dust_limit_satoshis: 354, max_htlc_value_in_flight_msat: 2100000, channel_reserve_satoshis: 1000, htlc_minimum_msat: 1, feerate_per_kw: 253, to_self_delay: 144, max_accepted_htlcs: 50, funding_pubkey: PublicKey(fcf5e3cd93e5bae4bd2865fb3b56fabcd3106c0dfa3345464b7a7edcfff3b7952da52785a989a3b41f39174695abd8d0a0c8fd8d918b51800cff24dca6a07185), revocation_basepoint: PublicKey(4808ab81a6b5906ef658fa2ac7e35706633226576621ba677bc7af0da199781dc005492318e33460e8bb2917c40c840b92f9159bba032ec3db5de9e9e25313cb), payment_point: PublicKey(2906c90ea2278eaaecd0e390756586092f6a7f30987b2cdf4148d11d8c77d6bf9b037ea0b46e7a5aa51c4d4955cb737d4fb6791d19ed44a07bb7ff3e48283d83), delayed_payment_basepoint: PublicKey(6c97115896bafd176f2257fcc841365465935125a7dfb241585b5c597e56d6a4bde7f7a9a3fcbce11532c301a2da5b5d32ff51d43389e4694900b5b52d41c648), htlc_basepoint: PublicKey(1d9f37ea983c531baf6ad5d751fead45f02db90388337b36ed2d5de3dee20dd658f2b61d6b9eba410c07409262513d4f1321c2af810f1032c1c97a8f661751b4), first_per_commitment_point: PublicKey(d02ec5434c50e92ebff5071c822bbbd4b789cc073dce9ffa5ac38dfcd6c802b9bf5f75b69fb49dbcae003bcb497aaa95febc5a59e5e2c2cc15bc3f07b549c181), channel_flags: 0, shutdown_scriptpubkey: Present(Script(OP_0 OP_PUSHBYTES_20 26f08713fdb565bbf22785fed26b46061e1ccda4)), channel_type: Some([0, 16]) }) from 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:59:32.392 TRACE [lightning_background_processor:520] Persisting ChannelManager...
2022-11-14 10:59:32.393 DEBUG [lightning::ln::peer_handler:1607] Handling SendAcceptChannel event in peer_handler for node 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d for channel c1522a9adcc3e6bffe5e9a1d5f25b62043cca731e9d8316ba458f5ce0a34a803
2022-11-14 10:59:32.393 TRACE [lightning::ln::peer_handler:980] Enqueueing message AcceptChannel { temporary_channel_id: [193, 82, 42, 154, 220, 195, 230, 191, 254, 94, 154, 29, 95, 37, 182, 32, 67, 204, 167, 49, 233, 216, 49, 107, 164, 88, 245, 206, 10, 52, 168, 3], dust_limit_satoshis: 354, max_htlc_value_in_flight_msat: 2100000, channel_reserve_satoshis: 1000, htlc_minimum_msat: 1, minimum_depth: 6, to_self_delay: 144, max_accepted_htlcs: 50, funding_pubkey: PublicKey(a198cffbebb1f0d1cc4ae5cb2e7ab9801d444eb33d3f2c1c117d7a64cb9a141c32638432f6215a77cd387bc398184a635378b7be203847f9250c39dec6b2907a), revocation_basepoint: PublicKey(03ac45ec454d8115ed6ad8ad16f9effc02cf43d3f599642fec027c566b253803513e6228ba84c43e7233f8ab37605f3c9b236228219d9155765a47c452728975), payment_point: PublicKey(ee4b03847ec5db4c20ad9782d9c6853e5d5308b513d6141845b628ec625770779fd903779f6a3c200dbc7683bf6b310a8077f69aa656be3406cf6e0172d2917c), delayed_payment_basepoint: PublicKey(bbcd6644ddd6036665cd44f8949e1a8f77ce25abaa5312f2965c922479a6c8d0abb07779469fc02b9ce419dcf0d88b3cabe5ffe9b1ce661afa629315a902c0b3), htlc_basepoint: PublicKey(e510e95e88c61ee35457221f41c5c07e120eccbd365a2f5181209b49159f889662697997d8d96c00084d5d93ca1bd2156eae74329ddaf3711df62b8960198347), first_per_commitment_point: PublicKey(5aa785eebba51c52432b87bd346da572655ea35a8a0a78764ef8919bd2580a8385edf3c50c1d7e1346b42c290730473a6f4761f1ca2d4e3dd44ae28d2da625ad), shutdown_scriptpubkey: Present(Script(OP_0 OP_PUSHBYTES_20 cca3c0ebc1d977a0d3eb2fe09fbdc8195151aa13)), channel_type: Some([0, 16]) } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:59:32.412 TRACE [lightning_background_processor:520] Done persisting ChannelManager.
2022-11-14 10:59:32.628 TRACE [lightning::ln::peer_handler:1283] Received message FundingCreated(FundingCreated { temporary_channel_id: [193, 82, 42, 154, 220, 195, 230, 191, 254, 94, 154, 29, 95, 37, 182, 32, 67, 204, 167, 49, 233, 216, 49, 107, 164, 88, 245, 206, 10, 52, 168, 3], funding_txid: 0e42605fe9a13d0d99865c1c8216827faa8dd1086a87e2921ecc9b5557161a37, funding_output_index: 0, signature: 304402204bd933fec4b4444cea8af6856e629b251adf4b4f7ad91f33c44907f2af9cf67902205ca635853623b1f6ea553180bef2b8696d54f971fff0300a2601428a1411f751 }) from 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
2022-11-14 10:59:32.650 TRACE [lightning::ln::channel:1448] Building commitment transaction number 281474976710655 (really 0 xor 261880908565150) for channel c1522a9adcc3e6bffe5e9a1d5f25b62043cca731e9d8316ba458f5ce0a34a803 for us, generated by remote with fee 253...
2022-11-14 10:59:32.651 TRACE [lightning::ln::channel:1614]    ...including to_remote output with value 20817
2022-11-14 10:59:32.651 TRACE [lightning::ln::channel:2151] Checking funding_created tx signature 4bd933fec4b4444cea8af6856e629b251adf4b4f7ad91f33c44907f2af9cf6795ca635853623b1f6ea553180bef2b8696d54f971fff0300a2601428a1411f751 by key 0395b7f3ffdc7e7a4b464533fa0d6c10d3bcfa563bfb6528bde4bae593cde3f5fc against tx 0200000001371a1657559bcc1e92e2876a08d18daa7f8216821c5c86990d3da1e95f60420e0000000000e62dee80015151000000000000160014c838471765a65d57180a51acfbd5edb02016ac039e4a8720 (sighash 9ce95aa7d161b2f16b45e47fdf5a23de7807ade1bf65dffd55a54aaecbae98ad) with redeemscript 475221021c149acb647a7d111c2c3f3db34e441d80b97a2ecbe54accd1f0b1ebfbcf98a1210395b7f3ffdc7e7a4b464533fa0d6c10d3bcfa563bfb6528bde4bae593cde3f5fc52ae for channel c1522a9adcc3e6bffe5e9a1d5f25b62043cca731e9d8316ba458f5ce0a34a803.
2022-11-14 10:59:32.652 TRACE [lightning::ln::channel:1448] Building commitment transaction number 281474976710655 (really 0 xor 261880908565150) for channel c1522a9adcc3e6bffe5e9a1d5f25b62043cca731e9d8316ba458f5ce0a34a803 for remote, generated by remote with fee 253...
2022-11-14 10:59:32.652 TRACE [lightning::ln::channel:1608]    ...including to_remote output with value 20817
2022-11-14 10:59:32.652 TRACE [lightning::ln::channel:2163] Initial counterparty tx for channel c1522a9adcc3e6bffe5e9a1d5f25b62043cca731e9d8316ba458f5ce0a34a803 is: txid 52af36e64348697a29bf2afcc5e8877b91e1dc5ec05f2ac7e8d82e997751b733 tx 0200000001371a1657559bcc1e92e2876a08d18daa7f8216821c5c86990d3da1e95f60420e0000000000e62dee800151510000000000002200207a9bf6584370a8007e2f0357119fcc6caaee4df78259e1555d11ab38d6be3cc79e4a8720
2022-11-14 10:59:32.652 TRACE [lightning::chain::channelmonitor:2092] Tracking new counterparty commitment transaction with txid 52af36e64348697a29bf2afcc5e8877b91e1dc5ec05f2ac7e8d82e997751b733 at commitment number 281474976710655 with 0 HTLC outputs
2022-11-14 10:59:32.652 INFO  [lightning::ln::channel:2247] Generated funding_signed for peer for channel 371a1657559bcc1e92e2876a08d18daa7f8216821c5c86990d3da1e95f60420e
2022-11-14 10:59:32.653 TRACE [lightning::chain::chainmonitor:583] Got new ChannelMonitor for channel 371a1657559bcc1e92e2876a08d18daa7f8216821c5c86990d3da1e95f60420e
2022-11-14 10:59:32.672 INFO  [lightning::chain::chainmonitor:597] Persistence of new ChannelMonitor for channel 371a1657559bcc1e92e2876a08d18daa7f8216821c5c86990d3da1e95f60420e completed
2022-11-14 10:59:32.673 DEBUG [lightning::ln::peer_handler:1628] Handling SendFundingSigned event in peer_handler for node 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d for channel 371a1657559bcc1e92e2876a08d18daa7f8216821c5c86990d3da1e95f60420e
2022-11-14 10:59:32.674 TRACE [lightning::ln::peer_handler:980] Enqueueing message FundingSigned { channel_id: [55, 26, 22, 87, 85, 155, 204, 30, 146, 226, 135, 106, 8, 209, 141, 170, 127, 130, 22, 130, 28, 92, 134, 153, 13, 61, 161, 233, 95, 96, 66, 14], signature: 3044022043aa9497aa7c1c20562c79ffc444a618b60c136b14cb237cbcc79b6b51f3469502206ae8499cbab8924de20c14eaae079cb3f61bdf9376d8ba6501e099388ce7bfe5 } to 0317673b5d263d92cc4e77989c37643f336a099c156181d6c9efb405e11d07276d
```

File: lib/main.dart [hunk 1]
```diff
@@ -152,13 +144,31 @@ class _TenTenOneState extends State<TenTenOneApp> {
...
+
+      final appSupportDir = await getApplicationSupportDirectory();
+      await api.initWallet(network: Network.Testnet, path: appSupportDir.path);
+
+      FLog.info(text: "Starting ldk node");
+      int listeningPort = 9735; // TODO: Move this to settings?
+      api
+          .runLdk(listeningPort: listeningPort)
+          .then((value) => FLog.info(text: "ldk node stopped."))
+          .catchError((error) => FLog.error(text: "ldk stopped with an error", exception: error));
```
Comment (bonomat) [inline lib/main.dart:159]: Do we even have to listen if no one can connect to us?

Comment (holzeis) [reply lib/main.dart]: no we don't 🙂

**Review by bonomat — APPROVED** [review]
LGTM

File: rust/src/api.rs [hunk 2]
```diff
@@ -69,7 +66,14 @@ pub fn get_address() -> Result<Address> {
 pub fn open_channel(peer_pubkey_and_ip_addr: String, channel_amount_sat: u64) -> Result<()> {
     let peer_info = parse_peer_info(peer_pubkey_and_ip_addr)?;
     let rt = Runtime::new()?;
-    rt.block_on(async { wallet::open_channel(peer_info, channel_amount_sat).await })
+    rt.block_on(async {
+        let _ = wallet::open_channel(peer_info, channel_amount_sat).await;
+        loop {
+            // looping here indefinitely to keep the connection with the maker alive.
+            std::thread::sleep(Duration::new(u64::MAX, u32::MAX));
```
Comment (bonomat) [inline rust/src/api.rs:73]: This will actually break, I've fixed it on my branch

Comment (holzeis) [reply rust/src/api.rs:73]: why?

**Pull request merged**